### PR TITLE
[rllib] add torch ppo

### LINKED ISSUE #3365: [rllib] PyTorch PPO / APPO
### Describe the problem
Currently we only have a simple pytorch example for A3C. It should be possible to "port" a bunch of pytorch algorithms onto the TorchPolicyGraph class and get them basically for free.

We'll also have to improve pytorch support in some ways (e.g., dict space support, lstm).

### PR DESCRIPTION
**Pull request by megankawakami**
## What do these changes do?

Add PyTorch support for PPOAgent.

## Related issue number

Closes https://github.com/ray-project/ray/issues/3365

### CODE REVIEW

File: python/ray/rllib/agents/ppo/torch_ppo_policy_graph.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        self.clip_param = clip_param
+        self.vf_clip_param = vf_clip_param
+        self.vf_loss_coeff = vf_loss_coeff
+        self.use_gae = use_gae
+
+    def forward(self, observations, value_target, advantages, actions, logits, vf_preds):
+    	def reduce_mean_valid(t):
+    		return torch.masked_select(t, valid_mask).sum(-1).sum()
+
+    	dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
```
Comment (richardliaw) [inline python/ray/rllib/agents/ppo/torch_ppo_policy_graph.py]:
```suggestion
    	 dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
```

File: python/ray/rllib/agents/ppo/torch_ppo_policy_graph.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        prev_dist = dist_cls(logits)
```
Comment (richardliaw) [inline python/ray/rllib/agents/ppo/torch_ppo_policy_graph.py]: why not use this to calculate entropy and KL?

**Review by richardliaw — COMMENTED** [review]
feel free to ping when this runs!

**Comment by ericl** [comment]
@megankawakami I ran pytorch: true | false with this command line and it seems the torch version doesn't learn as fast:

`./train.py --run=PPO --config='{"use_pytorch": {"grid_search": [true, false]}, "simple_optimizer": true, "num_workers": 0, "sample_batch_size": 100, "sgd_minibatch_size": 100, "train_batch_size": 100, "num_sgd_iter": 5, "log_level": "ERROR", "use_gae": false, "batch_mode": "complete_episodes"}' --env=CartPole-v0`

image

Could you give this a try? Also, I noticed that you aren't reporting kl or the loss metrics. You should override grad_info like in A2C to report these: https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/a3c/a3c_torch_policy_graph.py#L104 That way, it will be easier to debug why performance is lower by comparing these stats.

**Comment by ericl** [comment]
(auto-closing stale pr)

**Comment by jacekplocharczyk** [comment]
Hey,
what is going on with this PR?
I was thinking about adding PPO in Torch by myself but it's rather hard because I don't know Ray 😢
